en/guide/extending_csd.md

```diff
@@ -1,6 +1,121 @@
 # Extending Camera Streaming Daemon
 
-## Create a Custom Video Stream
+## Create a Custom Camera Device
+
+Camera-Streaming-Daemon has native support for auto detection of video cameras exported as Video4Linux(V4L2) devices. This covers most of regular Linux cameras, but there are some cameras that don't have V4L2 support or cases where user wants to set very specific parameters or do custom post-processing in video before exporting.
+
+For these use cases, it is possible to create a custom CameraDevice class, representing a single custom camera device.
+CameraServer class keeps a list of all created CameraComponents, each one holding a single available CameraDevice. These CameraDevice objects are used by CameraComponent class for camera control and ImageCapture and VideoStream classes to perform the image capture and streaming.
+
+[CameraDeviceGazebo](https://github.com/intel/camera-streaming-daemon/blob/master/src/CameraDeviceGazebo.cpp) is an example of a custom CameraDevice. We will explore it in order to learn all the steps necessary to add new custom CameraDevice in Camera-Streaming-Daemon.
+
+### CameraDevice
+
+CameraDevice is the abstract base class for all camera devices.
+
+```cpp
+class CameraDevice {
+```
+Every CameraDevice is identified by a unique string. For V4L2 class of camera devices, its the device node videox (For example video0 ). For Gazebo, its the string "gazebo".
+
+Based on how the camera frames can be read, there are two types of camera devices.
+- First is the camera device that supports gstreamer source element. The gst source element for such camera devices can be plugged in any gstreamer pipeline to implement the features like image capture, video streaming etc.
+- Second is the camera device that does not support gstreamer source element. Such type of class must implement the read function to return camera frame buffers. Gstreamer pipeline with appsrc element is created for such devices. The appsrc element is fed with frames returned by read function.
+
+### CameraParameters
+CameraParameters is the class to store and retrieve parameters and its value. Camera parameter is identified by a string and a corresponding integer ID. Parameter can be of any defined [types](http://mavlink.org/messages/common#MAV_PARAM_EXT_TYPE_UINT8). The parameter name, ID, type and its value(converted as string) is stored in the map data structure in the CameraParameters class. Few commonly used camera parameters are defined with name, ID and supported values. Nevertheless, developer can declare custom parameters as per their requirement.
```

Comment:
When you say "integer ID" do you mean "MAVLink type"? So a parameter is identified by a string and has a corresponding type?

"value(converted as string)" -> "value (converted to string)"

> Few commonly used camera ...

FMI Are you saying that this is rarely done, but you can if you want?

Reply:
> When you say "integer ID" do you mean "MAVLink type"? So a parameter is identified by a string and has a corresponding type?

No. For MAVLink, parameter name (string), type(int) and its value is what is needed. The ID is something that I have defined and is internal to CSD . Every parameter name is associated with an ID to enable usage of switch case in code.

> FMI Are you saying that this is rarely done, but you can if you want?

Not sure about the question. I wanted to define a very exhaustive list of camera parameters. So I added many in `CameraParameters` class. However its not exhaustive and users can add new parameters in new `CameraDeviceCustom` class

Reply:
OK, still confused. Mostly by ownership and content. What I think you are indicating is that:
1. Each CameraComponent has a CameraParameters to manage the parameters for the camera
2. This CameraParameters stores the cameras parameters as a map, with 
   * key being "string and id"
     * Id is to reference key in code - ie might be an enum
     * Why/when is the string needed.
   * The value being a string (or another value converted to a string)
   * Value has a MAVLink type. Is that type stored too?
3. The CameraCOmponent already defines a bunch of strings and IDs for common parameters.
   * What happens if a camera doesn't have a parameter that has already been defined?
4. You can add parameters to the CameraParameters in your device (or are you saying these get stored separate to CameraParameters)
5. Does this also store parameters from the camera's definition file? Or is this for only a particular type of parameter?

Reply:
> Why/when is the string needed.
MAVLink Parameter is identified with a name(string)

> Value has a MAVLink type. Is that type stored too?
Yes, need to send the type on GCS query

> What happens if a camera doesn't have a parameter that has already been defined?
CameraDevice will add set the value of the parameters supported, so unsupported parameters will not get added to the map.  Its like having a set of pre-defined param strings, pick those supported.

> You can add parameters to the CameraParameters in your device (or are you saying these get stored separate to CameraParameters)

There is one CameraParameters obj for every CameraComponent where all parameters related to `CameraDevice` , `ImageCapture` , `VideoCapture` and `VideoStream` will be stored.

> Does this also store parameters from the camera's definition file? Or is this for only a particular type of parameter?

Yes, whatever is there in camera def file has to be put in CameraParameters object. Whenever GCS does set/get, the values are set/get in/from CameraParameters obj.